transports/noise/src/io.rs

```diff
@@ -113,7 +107,7 @@ impl<T: AsyncWrite + Unpin> AsyncWrite for Output<T> {
         if this.send_offset == MAX_FRAME_LEN {
             trace!("write: sending {} bytes", MAX_FRAME_LEN);
             ready!(io.as_mut().poll_ready(cx))?;
-            io.as_mut().start_send(frame_buf)?;
+            io.as_mut().start_send(frame_buf.clone())?;
```

Comment:
This not good as we will be double-allocating each frame that we want to send. Can we somehow retain the use of a reference here?

Reply:
Yh. The clones made me doubt the viability of the PR but I opened it hoping to get some input. I don't think it's possible currently since `Framed` only implements `Sink` for `Encoder::Item`.

It seems to me that most encoding implementations don't need a `Self::Item` and really just use a `&Self::Item`(for example all the codecs provided by the `asynchronous_codec` follow this pattern). Another example is how the `serde::Serialize` trait takes a `&self`. Would it be viable to modify the `encode` fn definition?

Reply:
> Would it be viable to modify the `encode` fn definition?

Yeah! The upstream repository is here: https://github.com/mxinden/asynchronous-codec. I figured that we can solve this with GATs so I just went ahead and sent a PR to Max :)

https://github.com/mxinden/asynchronous-codec/pull/9

You can pull in a git dependency for the time being so you are unblocked :)

Reply:
Sweet! Nice solution. I'll go ahead and update the `Encode` impl on this end then